lib/history/sqlite.go

```diff
@@ -0,0 +1,379 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+	log "github.com/sirupsen/logrus"
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// clock is used to record event timestamps.
+	clock clockwork.Clock
+	// capacity specifies the max number of events that can be stored in the timeline.
+	capacity int
+	// size specifies the current number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sqlx.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *pb.SystemStatus
+	// mu locks timeline access.
+	mu sync.Mutex
+}
+
+// SQLiteTimelineConfig defines SQLiteTimeline configuration.
+type SQLiteTimelineConfig struct {
+	// DBPath specifies the database location.
+	DBPath string
+	// Capacity specifies the max number of events that can be stored in the timeline.
+	Capacity int
+	// Clock will be used to record event timestamps.
+	Clock clockwork.Clock
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified configuration.
+func NewSQLiteTimeline(config SQLiteTimelineConfig) (*SQLiteTimeline, error) {
+	database, err := initSQLite(config.DBPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		clock:    config.Clock,
+		capacity: config.Capacity,
+		size:     0,
+		database: database,
+		// TODO: store and recover lastStatus in case satellite agent restarts.
+		lastStatus: nil,
+	}, nil
+}
+
+// initSQLite initializes connection to database provided by dbPath and
+// initializes `events` table.
+func initSQLite(dbPath string) (*sqlx.DB, error) {
+	// TODO: Store sql db connection timeout as const
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```

Comment:
I would have this method (initSQLite) accept the context as a parameter and use it here.

I would even consider passing the context as a part of Config as well and use it as a parent context everywhere a context is needed. This way callers can choose to cancel if they choose to.

Reply:
Adding context to the config might make it confusing.
Should `RecordStatus` and `GetEvents` then be using the parent context stored in config? Is there a way we can enforce that? Or will this parent context only be used when initializing the SQLite database?

Reply:
Hmm, yeah it might get a little confusing if interface methods also accept contexts. In this case, maybe pass context to the constructor method and then, in turn, to the initSQLite?